### DISCUSSION TITLE
@Transactional(readOnly = True)  를 서비스 최상단으로 달아두고 싶어요.

### DISCUSSION THREAD

**Discussion by her0807** [woowacourse-teams/2022-ternoko#421] [Announcements]
## 문제 의식
@Transactional(readOnly = True) 는 사용을 하던 하지 않던 별 차이(사실 그동안 어떤 차이도 느끼지는 못했습니다.) 를 느끼지 못했기 때문에 
자주 사용하는 빈도와 크루들의 선호도에 따라서 취향 차이 정도가 있다고 생각하고 최초에 컨벤션을 정할 때  `@Transactional` 을 정했던 기억이 나요.
**그러던 중 동시성 이슈로 인해서 우연히 DB 에 락이 걸리는 상황** 을  겪으면서 Transactional 때문에 DB락이 걸릴 수도 있겠다는 생각이 들더라고요 ..?  😭

그래서 조금더 학습을 해봤습니다. 그랬더니 인프라 구조상 readOnly 가 주는 몇가지 이점을 발견했어요.

---

### @Transactional(readOnly = True) 로 변경해야하는 이유
1.  트랜잭션 인프라에서 몇 가지 최적화를 발생시킵니다. (FlushMode를 MANUAL로 설정하여 Persistence Provider 가 Entity Manager 를 닫을 때 **더티 체킹를 건너뛸 수 있도록 합니다** .). 
2.  사용하는 데이터베이스에 따라 테이블 잠금을 생략하거나 실수로 트리거할 수 있는 쓰기 작업을 거부하는 용도로 사용할 수도 있습니다. 
3. 일부 DB의 경우 master-slave 구조로 되어있다면 @Trasactional(readOnly = true) 를 사용하면 master DB를 읽는것이 아닌 읽기전용으로 slave DB를 읽게 설정할 수 있습니다. 즉, 상황에 따라 `workload` 분산을 해줄 수 있습니다.

---

## 요약
1. DB에 따라 @Transactional(readOnly=true) 를 했을 때 만약 update 나 insert 연산이 수행된다고 하여도 오류가 날 수 있고 나지 않을 수도 있습니다.
2. @Transactional(readOnly=true) 를 하면 더티체킹을 건너뛰는등 여러 최적화 작업이 일어나기 때문에 성능이 향상됩니다.
3. DB가 master-slave 구조로 되어있다면 추가 설정을 통해서 CUD(create, update, delete) 는 master DB 로 가게 설정하고 Read 는 slave 로 가게 설정하여 workload를 분산할 수 있습니다.

### ✏️ 결론 :  @Transactional(readOnly = True)  를 서비스 최상단으로 달아두고 싶어요. 😀

---

## 참고 자료
- https://vladmihalcea.com/spring-read-only-transaction-hibernate-optimization/
- https://stackoverflow.com/questions/44984781/what-are-advantages-of-using-transactionalreadonly-true
- https://algopoolja.tistory.com/104

**Comment by soominsohn** [comment]
수달 좋은 자료 공유 고마워요!
수달 말대로 읽기 전용 쿼리만 사용하는 서비스는 @transactional(readOnly = true)를 붙이는게 좋겠네요.
그러나 저는 서비스 최상단에는 readOnly = true를 붙이는게 아닌 @transactional 어노테이션만 붙이는게 좋다고 생각해요.

만일 수달이 언급하신 요약 3번 대로 `DB가 master-slave 구조로 되어있다면 추가 설정을 통해서 CUD(create, update, delete) 는 master DB 로 가게 설정하고 Read 는 slave 로 가게 설정하여 workload를 분산할 수 있습니다.` 
이 이루어진다면
서비스 최상단에 @transactional(readOnly = true)를 붙이고, CUD가 일어나는 메서드에 @transactional 을 따로 붙이는 행위를 하지 않는다면 다음과 같은 문제가 생길 것 같네요. https://findmypiece.tistory.com/259

제가 언급한 내용은 혹시 기억하실지 모르겠지만 수달이 테코톡 주제 질문하러 토미께 갔을 때 토미가 수달, 저, 애쉬한테 해주신 말씀이에요! (이 부분은 확인하면 지울게요)

그래서 제 결론은
서비스 최상단에는 @transactional 어노테이션만 사용하고, readOnly 옵션이 필요한 메서드에만 옵션을 달자입니다 😄

**Comment by dongho108** [comment]
저희 readOnly 쓰고 있지 않나요?
![스크린샷 2022-09-29 12 28 53](https://user-images.githubusercontent.com/54317630/192932081-c47e4c7e-111e-47f9-b074-5dbe4422901f.png)

**Comment by dongho108** [comment]
휴먼에러방지를 위해 현행유지 땅땅땅
